File: c/test/tlsprobe.c
/* One HTTPS round-trip, and nothing else.
 *
 * test/tlscheck.sh drives this against a local TLS server to prove the
 * four obligations a TLS binding in this repository has: chain
 * verification, HOSTNAME verification, SNI, and additive extra roots from
 * SEKRETO_CA_BUNDLE.
 *
 * The hostname half needs its own proof and cannot borrow anyone else's:
 * `make integration` contains no https URL at all, and `test/realstores.sh`
 * has no NEGATIVE hostname case - so nothing in either suite would notice
 * a port that verified the chain and then accepted a certificate issued
 * for somebody else. That case is number 4 in tlscheck.sh.
 *
 * Prints one line - `status=<code>` or `error=<message>` - so the shell
 * driver can assert on it.
 */

#define _POSIX_C_SOURCE 200809L

#include <stdio.h>

#include "sekreto.h"

int main(int argc, char **argv) {
  sek_pool *pool = sek_pool_new();
  int status = 0;
  char *body = NULL;
  sek_err err;

  if (2 > argc) {
    printf("error=usage: tlsprobe <url>\n");
    sek_pool_free(pool);
    return 2;
  }

  err = sek_fetch(pool, "GET", argv[1], NULL, NULL, &status, &body);

  if (NULL != err) {
    printf("error=%s\n", err);
    sek_pool_free(pool);
    return 1;
  }

  printf("status=%d\n", status);

  sek_pool_free(pool);

  return 0;
}

File: lua/test/tlsprobe.lua
-- One TLS handshake, and whether it was accepted.
--
-- Driven by test/tlscheck.sh. Prints `OK` when the handshake completed
-- and the peer was verified, or `REFUSED <reason>` when it was not.
--
-- Usage: lua5.4 test/tlsprobe.lua <host> <port>

package.path = 'src/?.lua;' .. package.path

local net = require('sekreto.net')

local host = arg[1]
local port = tonumber(arg[2])

local request = 'GET / HTTP/1.1\r\nHost: ' .. host ..
  '\r\nAccept: */*\r\nConnection: close\r\n\r\n'

local got, why = net.fetch(host, port, true, request, 10000, 8 * 1024 * 1024)

if nil == got then
  print('REFUSED ' .. why)
  os.exit(0)
end

print('OK')
